### PR TITLE
feat(zap): silent zap logging

### LINKED ISSUE #311: Add ApplicationOption to specify logger (and others?)
Like `WithCoreLogger`, which replaces zap.Module in https://github.com/i-love-flamingo/flamingo/blob/5909a729b3075163e0e961f702de47a482a6dfbb/app.go#L131

### PR DESCRIPTION
**Pull request by IvanMaidurov**
Logger will write to output only if error occurred during request

Closes #311

### CODE REVIEW

**Comment by IvanMaidurov** [comment]
easy test with no parallel requests but with parallel writings:

```
name              old time/op    new time/op    delta
Log/spam_test-10     2.28s ± 1%     0.45s ±77%   -80.29%  (p=0.000 n=10+10)

name              old alloc/op   new alloc/op   delta
Log/spam_test-10    95.0MB ± 1%   216.7MB ± 0%  +128.20%  (p=0.000 n=9+10)

name              old allocs/op  new allocs/op  delta
Log/spam_test-10     1.33M ± 1%     1.39M ± 1%    +4.43%  (p=0.000 n=10+10)
```

100 parallel requests, each. with 300 parallel writes:

```
name              old time/op    new time/op    delta
Log/spam_test-10     2.02s ± 1%     2.13s ± 1%     +5.36%  (p=0.000 n=10+10)

name              old alloc/op   new alloc/op   delta
Log/spam_test-10     198MB ± 1%   11409MB ± 0%  +5657.54%  (p=0.000 n=10+9)

name              old allocs/op  new allocs/op  delta
Log/spam_test-10     3.20M ± 0%     6.07M ± 0%    +89.84%  (p=0.000 n=10+10)
```

File: app.go [hunk 1]
```diff
@@ -128,7 +129,7 @@ func NewApplication(modules []dingo.Module, options ...ApplicationOption) (*Appl
 
 	modules = append([]dingo.Module{
 		new(framework.InitModule),
-		new(zap.Module),
```
Comment (bastianccm) [inline app.go:131]: I would not change the default here, but instead do that with #311

File: core/silentlogger/module.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	}
+)
+
+// Configure the ZAP logger as flamingo.Logger (in JSON mode kibana compatible)
+func (m *Module) Configure(injector *dingo.Injector) {
+	registry := new(LoggingContextRegistry)
+
+	injector.Bind(registry).AsEagerSingleton()
+
+	injector.Bind(new(flamingo.Logger)).ToProvider(getSilentLogger)
```
Comment (bastianccm) [inline core/silentlogger/module.go:26]: `getSilentLogger` is quite a "heavy" implementation, can this be a performance issue if we call the factory for every logger instance?

File: core/silentlogger/module.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	Module struct{}
+
+	shutdownEventSubscriber struct {
+		logger flamingo.Logger
+	}
+)
+
+// Configure the ZAP logger as flamingo.Logger (in JSON mode kibana compatible)
+func (m *Module) Configure(injector *dingo.Injector) {
+	registry := new(LoggingContextRegistry)
```
Comment (bastianccm) [inline core/silentlogger/module.go:22]: This is not necessary, rather reference them in the bindings `Bind(new(LoggingContextRegistry))`...

**Review by bastianccm — CHANGES_REQUESTED** [review]

File: core/silentlogger/module.go (hunk 3 above)
Comment (IvanMaidurov) [reply core/silentlogger/module.go:22]: in line 27 I used this reference in ```flamingo.BindEventSubscriber(injector).To(registry)```, will EagerSingleton still be referenced if I create two instances?

File: core/silentlogger/module.go (hunk 2 above)
Comment (IvanMaidurov) [reply core/silentlogger/module.go:26]: I guess no really big differences, all info is getting cleared every request end. So it either lives in one instance, or spread across different instances. By itself it's not heavy in my opinion

**Review by bastianccm — APPROVED** [review]

**Pull request merged**